### ISSUE TITLE
Scratch-Style JavaScript Code Editor

### ISSUE THREAD

**Issue by PullJosh** [PullJosh/leopard-website#38]
Hey y'all! After hearing about my little sister's experience using Leopard for the first time, it seems like there are some opportunities to make the end-user experience better. In particular, the translation for her project worked great, but then she was kinda stuck. Immediate roadblocks were:
- How do I publish my Leopard project? ([These instructions](https://leopardjs.com/help/publish-with-vercel) were too complicated)
- How do I learn JavaScript/how do I make changes?

There are lots of possible solutions centered around creating better **educational resources**, and I'd like to pursue some of those myself, but there are also some **user experience** issues that are just fundamentally _harder than they need to be_. I'm okay with things being hard if it's a necessary difficulty that helps people learn, but if our goal is to be an educational resource to help people transition from Scratch to JS/web development, I think we should remove as many of the _unnecessary_ hurdles as possible.

**To that end, I've been experimenting with a custom editor that could be an option alongside the Codesandbox route.** (To be honest, I'd like to have lots of export targets, such as Replit, Codepen, a simple [.zip](https://github.com/PullJosh/leopard-website/issues/33), etc, with this custom editor being just one of them–albeit maybe the default one if it becomes good enough at some point in the future.)

<img width="1384" alt="Screenshot of code editor" src="https://github.com/PullJosh/leopard-website/assets/5458180/8da83100-67ff-4a7e-befc-7269ce1a16ff">

<img width="1384" alt="Screenshot of costume preview" src="https://github.com/PullJosh/leopard-website/assets/5458180/b7638275-6edf-42f3-949e-88fdb76edae8">

My goal with the design is to be a fully-featured editor which allows users to edit **any static website** any way they want. However, I also want the editor to be extremely reminiscent of the Scratch editor. For that reason, rather than exposing the files as a simple file tree on the left, I'm putting the top-level folder in the bottom right as a grid of icons (which essentially serves as a "sprite picker" for conventionally-organized Leopard projects), exposing the contents of the selected folder as a list of tabs, and then if you choose a tab that is a folder (such as "costumes" or "sounds"), exposing the contents of that directory as a file tree on the left–equivalent to the list of costumes/sounds in Scratch.

In all, this allows users to organize their files in any way they want, because _it's just a file system_, but to get the full Scratch editor feeling along the way.

**Comment by sfederici** [comment]
What a wonderful idea! Is there an alpha/beta version available to be tested?

**Comment by PullJosh** [comment]
> What a wonderful idea! Is there an alpha/beta version available to be tested?

Thank you! I've been doing all the development locally so far so I don't have a publicly testable version.

I am currently very busy doing my student teaching in a high school math class, so I can't put it online right now. But I'll be done with lead teaching in just 10 days, on March 22nd, so I'll have way more programming time after that and will make this a priority. 😄

**Comment by sfederici** [comment]
Excellent! Let me know when test time is on

**Comment by PullJosh** [comment]
The editor is **extremely** not ready for prime-time nor even for outside contributors because so much still needs to be ripped out and changed. (The todo list lives in my head, mostly.) But for the sake of openness, I did publish my current code to the `project-editor` branch. (Please nobody make any PRs yet–it's gotta become a bit more stable before it's ready for lots of outside input.)

**Comment by sfederici** [comment]
> The editor is **extremely** not ready for prime-time nor even for outside contributors because so much still needs to be ripped out and changed. (The todo list lives in my head, mostly.) But for the sake of openness, I did publish my current code to the `project-editor` branch. (Please nobody make any PRs yet–it's gotta become a bit more stable before it's ready for lots of outside input.)

I'm eager to try it! I did an `npm ci` and got good results. But then my `npm start` raised an error. I guess I will have to wait a little more

**Comment by cristian-spiescu** [comment]
Hi guys.

FYI, I'm working w/ some colleagues on a VS code extension that allows editing scratch + js.

**Comment by cristian-spiescu** [comment]
> Is this a Scratch/Leopard integration?

Yes indeed. 

I'm playing together w/ my kids w/ Scratch. Scratch is a wonderful tool for the kids. I am using it to create small programming challenges. But in order to do this, a bit of "framework" code is needed (i.e. my job in the project). And for me, writing in Scratch is not so awesome + it has a lot of limitations. That's how I found out about `leopard-js`. However, even if I write the majority of my contribution in JS, I still need the Scratch editor to quickly verify how my code combines w/ Scratch. Hence my idea of VS Code extension.

**Comment by sfederici** [comment]
I like the idea a lot. So, please, keep me posted. I'm available for alpha testing too.

**Comment by cristian-spiescu** [comment]
@PullJosh, thanks.

The leopard repo has "discussions" (forum). In my opinion, this is the correct approach. If contributors (e.g. myself) want to showcase something, then they create a post there. I see there is already a [dedicated section](https://github.com/leopard-js/leopard/discussions/categories/show-and-tell). 

I appreciate Discord for its real-time communication capabilities. In my opinion, for dev projects, sometimes it helps, sometimes it distracts. But is a matter of taste. I think that forum (or Stack Overflow) style messages force people to write in a structured way, and to receive well articulated answers. With Discord/chat style messages: the (technical) communication is more chaotic.

**Comment by PullJosh** [comment]
@cristian-spiescu That's true. I much prefer the forum format as well. I've worried that Discord is where most people are, and we have to meet them there, but perhaps developers are happy living on Github. If you'd like to create a discussion about your project in the Leopard repo, that would be awesome. 👍

**Comment by cristian-spiescu** [comment]
> create a discussion about your project in the Leopard repo

Done [here](https://github.com/leopard-js/leopard/discussions/194).

**Comment by PullJosh** [comment]
**Exciting update! 🎉** For all its many faults, the test website is officially online:
https://leopard-website-git-project-editor-josh-pullens-projects.vercel.app/

Don't mind the long URL. Give it a try and let me know what you think! Keep in mind that all your projects could be deleted at any time, so don't create anything precious. Also, assume my security is terrible just in case.

**Comment by PullJosh** [comment]
I just deleted the alpha accounts and projects in a big batch of changes that would have been challenging to perform otherwise–hopefully everyone heeded the "Nothing you create here is private, and it may be deleted at any time" warning. Feel free to re-register and keep testing.

---

Since the beginning of alpha testing, there have been a bunch of changes and improvements. A few highlights:
- It is now possible to upload assets to your project
- Projects are now private by default and can be shared/unshared
- There is now a "my stuff" page just like Scratch for viewing all your projects, including the unshared ones
- The project preview page now has a green flag and stop sign
- The homepage now has a better UI for converting projects, including the option to generate a blank project
- The homepage convert box now allows users to choose between the Leopard editor, CodeSandbox, or downloading as a zip
- User accounts now require an email address when registering. I would prefer it if this wasn't needed, but for the sake of spam prevention and allowing password resets, I don't see any better solution. I am still working on a process for allowing users to confirm their email address (and likely requiring this before allowing users to share projects).

A few key things that still need to be done before this can be released to the public:
- [ ] Project size limits (maximum MB per user or something like that)
- [ ] Moderation tools (the ability to ban users, etc)
- [ ] Make it possible to delete files in the editor (and ideally rename files as well, although users could work around this limitation)
- [ ] Serve project previews from a separate domain for the sake of security. Right now projects could call the leopard website's APIs on the user's behalf, because that user's cookie would be included in the request.

There is a never-ending list of additional improvements I'd like to make, but the above list seems like the essential few before a public release is a net good for users rather than a net bad.

---

@sfederici Pinging you because you were curious about this and I'm not sure if you've seen it yet.

**Comment by PullJosh** [comment]
**tl;dr:** Made some security updates. Domains had to change. Go to https://mrpullen.com/ and sign in with your old username and password. Yes, it's a silly choice of domain.

---

**Okay. So. This is a little goofy.** Project previews are now served from a subdomain. This is much, much better for security. (For example, we really don't want projects to have access to your Leopard session cookie.)
- **Tired 😱:** `https://leopardjs.com/api/preview/YOUR_PROJECT_ID/`
- **Wired 👍:** `https://YOUR_PROJECT_ID.preview.leopardjs.com`

However, due to a bunch of unimportant details, this subdomain feature can only be implemented if you are using a proper pay-$12-a-year domain name, which means the old long-as-crap Vercel subdomain no longer works.

So we need a new staging domain, but I'm too cheap to buy one. Fortunately, I own exactly one .com domain that is not currently in use! **For that reason, the new Leopard editor staging site now lives at https://mrpullen.com/** 🤦‍♂️

![It's free real estate gif](https://media4.giphy.com/media/8VSaCyIdcnbuE/200w.gif?cid=6c09b952gtnedycjr6j8k9cg5rvm1tt9y45ackcwvnsfkzh9&ep=v1_gifs_search&rid=200w.gif&ct=g)

**Comment by sfederici** [comment]
> @sfederici Pinging you because you were curious about this and I'm not sure if you've seen it yet.

Hi Josh. Sorry for disappearing. Lessons, exams and conferences were a bit too much in the last weeks. I tried https://mrpullen.com/ just now and I got

404: NOT_FOUND Code: DEPLOYMENT_NOT_FOUND ID: fra1::xz76n-1719845979957-2b1ac7d42207

**Comment by PullJosh** [comment]
@sfederici You could not have popped it at a more perfect time. I just took down the staging site because **the new version of the website, including the new editor, officially launched on the production website, [27 minutes ago](https://github.com/PullJosh/leopard-website/pull/44)! 🎉**

I am still working on writing up little posts to announce the new update, but in the meantime it would be lovely to have your feedback. There's still a ton to improve, but the main thing to look out for right now is just bugs or things that are completely broken.

**Comment by PullJosh** [comment]
For anyone who is interested:
- Announcement forum post: https://scratch.mit.edu/discuss/topic/420162/?page=53#post-8036132
- Announcement video: https://www.youtube.com/watch?v=23tB-1HWPnM
